clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7443,6 +7443,26 @@ static bool checkForDuplicateAttribute(Sema &S, Decl *D,
   return false;
 }
 
+// Checks if FPGA memory attributes apply on valid variables.
+// Returns true if an error occured.
+static bool CheckValidFPGAMemoryAttributesVar(Sema &S, Decl *D,
+                                              const AttributeCommonInfo &CI) {
+  if (const auto *VD = dyn_cast<VarDecl>(D)){
+    if (!(VD->getKind() != Decl::ImplicitParam &&
+          VD->getKind() != Decl::NonTypeTemplateParm &&
+          ((VD->getStorageClass() == SC_Static || VD->hasLocalStorage()) ||
+           (VD->getKind() != Decl::ParmVar &&
+            (S.isTypeDecoratedWithDeclAttribute<SYCLDeviceGlobalAttr>(VD->getType())
+             || VD->getType().isConstQualified()
+             || VD->getType().getAddressSpace() ==
+                    LangAS::opencl_constant))))) {
```

Comment:
Treat this as purely a suggestion, but I am just confused by the logic of the if statement.   I think what I have below is equivalent and reads better to me.

```suggestion
    if (!(VD->getKind() != Decl::ImplicitParam &&
          VD->getKind() != Decl::NonTypeTemplateParm &&
          VD->getKind() != Decl::ParmVar &&
          (VD->getStorageClass() == SC_Static || VD->hasLocalStorage() ||
            S.isTypeDecoratedWithDeclAttribute<SYCLDeviceGlobalAttr>(VD->getType() ||
            VD->getType().isConstQualified() ||
            VD->getType().getAddressSpace() ==
                    LangAS::opencl_constant))) {
```

Reply:
Also, in the refactoring, where do we check that it is a Field?  I am just curious.

Reply:
Thanks @premanandrao for feedbacks. I have updated logic and also added a check for Field.